### PR TITLE
Don't pick .lock'd and new dirs for deletion when problem dirs grow over size limit

### PR DESCRIPTION
**Pull request by michalfabik**
See commits.

Resolves https://github.com/abrt/abrt/issues/1450.  
  
See also https://github.com/abrt/abrt/pull/1481

### CODE REVIEW

File: src/lib/dirsize.c [hunk 1]
```diff
@@ -130,9 +131,12 @@ double libreport_get_dirsize_find_largest_dir(
                     }
                     else
                     {
-                        maxsz = sz;
-                        free(*worst_dir);
-                        *worst_dir = libreport_xstrdup(ep->d_name);
+                        if (strcmp(ep->d_name, proc_dir))
```
Comment (mgrabovsky) [inline src/lib/dirsize.c:134]: This segfaults if `proc_dir == NULL`.

Also, I'd prefer to have the explicit `!= 0` there.

File: src/lib/dirsize.c [hunk 2]
```diff
@@ -96,6 +97,13 @@ double libreport_get_dirsize_find_largest_dir(
             log_debug("Skipping %s': sosreport is being generated.", dname);
             goto next;
         }
+        if (lstat(libreport_concat_path_file(dname, ".lock"), &statbuf) == 0)
```
Comment (ernestask) [inline src/lib/dirsize.c:100]: Aren’t you leaking the string here?

File: src/lib/dirsize.c [hunk 3]
```diff
@@ -96,6 +97,13 @@ double libreport_get_dirsize_find_largest_dir(
...
+        {
+            log_warning("Skipping %s: directory locked. Is a backtrace being generated?", dname);
+            double sz = libreport_get_dirsize(dname);
```
Comment (ernestask) [inline src/lib/dirsize.c:103]: I’m not sure the intermediate variable is really necessary in this case.

File: src/lib/dirsize.c [hunk 4]
```diff
@@ -96,6 +97,13 @@ double libreport_get_dirsize_find_largest_dir(
...
+            size += sz;
```
Comment (ernestask) [inline src/lib/dirsize.c:104]: Why is this needed if we’re skipping the directory? I’m looking at the sosreport-still-being-generated case above and it doesn’t do that.

Comment (michalfabik) [reply src/lib/dirsize.c:104]: I thought the directory's size still counts to the total size of the directory where problem directories are stored `/var/spool/abrt`, even if we're skipping it for now.

Comment (ernestask) [reply src/lib/dirsize.c:104]: Yeah, that makes sense. Do you think we should do the same above, then?

Comment (ernestask) [reply src/lib/dirsize.c:104]: *Ideally*, we would fold the two cases into yours, but that might be more involved, given that sosreport generation comes from an event handler.

File: src/lib/dirsize.c (hunk 2 above)
Comment (ernestask) [reply src/lib/dirsize.c:100]: Could we also use `g_build_filename()` here?

File: src/lib/dirsize.c (hunk 4 above)
Comment (michalfabik) [reply src/lib/dirsize.c:104]: > Yeah, that makes sense. Do you think we should do the same above, then?  
  
Yeah, makes sense to me.

File: src/lib/dirsize.c (hunk 2 above)
Comment (michalfabik) [reply src/lib/dirsize.c:100]: > Aren’t you leaking the string here?  
  
`dname`? That one is `free()`'d at `next:`.

Comment (michalfabik) [reply src/lib/dirsize.c:100]: > Could we also use `g_build_filename()` here?  
  
Oh, absolutely, but I'd rather do it in a separate pull request along with all the other GLib migration changes.

Comment (ernestask) [reply src/lib/dirsize.c:100]: > Oh, absolutely, but I'd rather do it in a separate pull request along with all the other GLib migration changes.

Okay, just throwing it out there to make it one line less when you do that.

Comment (ernestask) [reply src/lib/dirsize.c:100]: > `dname`? That one is `free()`'d at `next:`.

`libreport_concat_path_file()` returns a heap-allocated string, which should be freed.

Comment (michalfabik) [reply src/lib/dirsize.c]: Leaks fixed.

File: src/lib/dirsize.c [hunk 5]
```diff
@@ -147,6 +149,8 @@ double libreport_get_dirsize_find_largest_dir(
         }
  next:
         free(dname);
+        free(sosreport_path);
+        free(lock_path);
```
Comment (mgrabovsky) [inline src/lib/dirsize.c:153]: I'd recommend to use `g_autofree` instead. (Thanks, Ernestas.)

Comment (michalfabik) [reply src/lib/dirsize.c:153]: Right, I actually thought about this right after I pushed the commit. :\

Comment (michalfabik) [reply src/lib/dirsize.c]: Like this? And we got rid of the `goto`, too.

Comment (mgrabovsky) [reply src/lib/dirsize.c]: Yup, looks better now :+1:

File: src/lib/dirsize.c [hunk 6]
```diff
@@ -123,9 +133,12 @@ double libreport_get_dirsize_find_largest_dir(
                     }
                     else
                     {
-                        maxsz = sz;
-                        free(*worst_dir);
-                        *worst_dir = libreport_xstrdup(ep->d_name);
+                        if (!proc_dir || strcmp(proc_dir, ep->d_name) != 0)
```
Comment (ernestask) [inline src/lib/dirsize.c:136]: You could fold this into an `else if` and save some horizontal whitespace.

**Comment by ernestask** [comment]
Looks good to me overall.
